akka-sample-persistence-dc-scala/src/main/scala/sample/persistence/res/auction/Auction.scala

```diff
@@ -0,0 +1,243 @@
+package sample.persistence.res.auction
+
+import java.time.Instant
+
+import akka.actor.typed.{ActorRef, Behavior}
+import akka.actor.typed.scaladsl.{ActorContext, Behaviors, LoggerOps, TimerScheduler}
+import akka.persistence.cassandra.query.javadsl.CassandraReadJournal
+import akka.persistence.typed.{RecoveryCompleted, ReplicaId, ReplicationId}
+import akka.persistence.typed.scaladsl.{Effect, EventSourcedBehavior, ReplicatedEventSourcing, ReplicationContext}
+import sample.persistence.res.{CborSerializable, MainApp}
+
+import scala.concurrent.duration._
+
+class Auction(
+                     context: ActorContext[Auction.Command],
+                     replicationContext: ReplicationContext,
+                     timers: TimerScheduler[Auction.Command],
+                     closingAt: Instant,
+                     responsibleForClosing: Boolean) {
+  import Auction._
+
+  private def behavior(initialBid: Bid): EventSourcedBehavior[Command, Event, AuctionState] =
+    EventSourcedBehavior(
+      replicationContext.persistenceId,
+      AuctionState(phase = Running, highestBid = initialBid, highestCounterOffer = initialBid.offer),
+      commandHandler,
+      eventHandler).receiveSignal {
+      case (state, RecoveryCompleted) => recoveryCompleted(state)
+    }
+
+  private def recoveryCompleted(state: AuctionState): Unit = {
+    if (shouldClose(state))
+      context.self ! Close
+
+    val millisUntilClosing = closingAt.toEpochMilli - replicationContext.currentTimeMillis()
+    timers.startSingleTimer(Finish, millisUntilClosing.millis)
+  }
+
+  private def commandHandler(state: AuctionState, command: Command): Effect[Event, AuctionState] = {
+    state.phase match {
+      case Closing(_) | Closed =>
+        command match {
+          case GetHighestBid(replyTo) =>
+            replyTo ! state.highestBid.copy(offer = state.highestCounterOffer) // TODO this is not as described
```

Comment:
nitpicking/style rant ahead...

I find that each time we have nested pattern matching the code becomes less easy to follow as we add a mental context switch. We need then to parse the different levels to identify what belongs to what.

We could instead have...

```scala
private def closingCommandHandler(state: Closing, command: Command): Effect[Event, AuctionState]
private def runningCommandHandler(state: Running, command: Command): Effect[Event, AuctionState]
```

and then pattern match on the state higher up when passing it to the behavior: 

```scala
(state, command) =>
  state match {
    case closing: Closing => closingCommandHandler(closing, command)
    case running: Running => runningCommandHandler(running, command)
  },
```

Reply:
Correction: 

```scala
private def readOnlyCommandHandler(state: AuctionState, command: Command): Effect[Event, AuctionState]
private def runningCommandHandler(state: Running, command: Command): Effect[Event, AuctionState]
```

```scala
(state, command) =>
  state match {
    case Closing(_) | Closed  => readOnlyCommandHandler(state, command)
    case running: Running     => runningCommandHandler(running, command)
  },
```

Reply:
Looks good, I'll change it